Update to return correct value for multiple_domains().

ghost opened the pull request:
requests/cookies.py

Adjust flow to add domains to list as the first operation in the loop, then convert to a set and check the length of that set is greater than 1.

Add tests to pass for a single domain, two cookies with the same domain, and multiple domains.

Perhaps not the most efficient method, but it works.

ghost commented:
#4557

nateprewitt commented on requests/cookies.py:
```diff
@@ -292,9 +292,11 @@ def multiple_domains(self):
-            if cookie.domain is not None and cookie.domain in domains:
-                return True
             domains.append(cookie.domain)
+            domains_set = set(domains)
```
A few quick notes here.

We’re rebuilding the set every loop which probably isn’t necessary. Since it’s serving the same purpose of `domains`, what do you think about making `domains` a set?

The other issue I’m seeing is the conditional. Let’s take an example set of domains `['google.com', None, 'google.com']`. The first two loops will add both 'google.com' and None to `domains` and the third loop will satisfy the criteria despite there only being one domain. It may be better if we only add if the cookie.domain is not None and then have the single check for `len(domains)` being > 1

gahancorpcfo replied:
I think you're correct there, I'll update the pr tomorrow.

ghost commented:
@nateprewitt Thank you kindly for the helpful feedback!

I've adjusted the method to make use of set in place of the domains list.  I've also adjusted the structure of the conditional a bit and added an additional test check that null values are handled correctly by the method.

nateprewitt commented on tests/test_requests.py:
```diff
@@ -385,6 +385,27 @@ def test_param_cookiejar_works(self, httpbin):
+
+        assert r.cookies.multiple_domains() == False
+
+        r.cookies.set('baz', 'foo', domain='.google.com')
```
If we’re going to add this to the test let’s verify it’s working as expected. 

We can probably collapse a few of the new lines too if you don’t think it affects readability. Perhaps no extra newline between the cookies.set and assert lines?

gahancorpcfo replied:
Is it worth a new test method to make testing the ['.google.com',None,'.google.com'] case a bit less awkward?

nateprewitt replied:
If you think that makes it easier to grock, I don’t have an issue with it. I’d encourage a glance at pytest’s `parameterize` decorator which will allow you to run the same test with different input.

ghost commented:
@nateprewitt You're very welcome, sir.

The other updates you requested have been made.

nateprewitt commented:
@gahancorpcfo, It looks like you may have accidentally merged master into your commit. Would you mind only including the changes related to this PR. The easiest way to do this is likely to reset HEAD~1 and then `git add -p your-relevant-files` or cherry pick from another branch if you have it.

ghost commented:
I believe I've managed to produce a single commit with the relevant changes that was not merged with master. 

. . . I hope. 

@nateprewitt Thank you very much for your help with this.

nateprewitt commented:
@gahancorpcfo it looks like you still have some extra code in cookies.py. Would you mind removing that? `git add -p` will allow you to easily select the chunks you need.

ghost commented:
@nateprewitt 

The extra code has been removed from the commit now, it should be good to go.

nateprewitt commented:
@gahancorpcfo thanks so much for seeing this through! We have an issue right now with our Windows build environment that's failing on Python 3.4, so I'm going to wait to merge this. We'll get this in once that's resolved.

As for the issue, it looks like https://github.com/jonparrott/cmarkgfm has had issues with Windows wheel builds recently between versions 0.2.0 and 0.4.0. I'll need to reach out to Jon to see if this is supposed to be fixed. That is unless someone else would like to spend some time digging and opening a relevant issue with the repository.

nateprewitt commented:
Just as a follow up note, jonparrott/cmarkgfm#2 is the issue we're hitting right now. Until a wheel is built for 3.4 (hopefully in the next week), 3.4 won't be passing on Windows under appveyor.

ghost commented:
@nateprewitt Thanks for the heads up!  I'm happy to help.

theacodes commented:
Sorry it took so long, but 3.4 wheels should now be available for Windows. :)

nateprewitt commented:
No worries, thanks @jonparrott!

sethmlarson commented:
In an effort to clean up the issue tracker to only have issues that are still relevant to the project we've done a quick pass and decided this issue may no longer be relevant for a variety of potential reasons, including:

* Applies to a much older version, unclear whether the issue still applies.
* Change requires a backwards incompatible release and it's unclear if the benefits are worth the migration effort from the community.
* There isn't a clear demand from the community on the change landing in Requests.

If you think the issue should remain open, please comment so below or open a new issue and link back to the original issue. Again, thank you for opening the issue and for the discussion, it's much appreciated.
